### PR TITLE
VTK-m: Add testing variant.

### PR DESCRIPTION
**Pull request by brugger1**
I added a `testing` variant to the VTK-m package. If enabled it sets `VTKm_ENABLE_TESTING_LIBRARY:BOOL=ON` otherwise it sets it to `OFF`.

I tested `spack install vtk-m` and it didn't install any testing files and then I tested `spack install vtk-m+testing` and it installed the testing header files as shown below.
```
find . -print | grep testing
./vtkm/testing
./vtkm/testing/Testing.h
./vtkm/testing/VecTraitsTests.h
./vtkm/cont/testing
./vtkm/cont/testing/vtkm_cont_testing_export.h
./vtkm/cont/testing/ExplicitTestData.h
./vtkm/cont/testing/MakeTestDataSet.h
./vtkm/cont/testing/Testing.h
./vtkm/cont/testing/TestingArrayHandles.h
./vtkm/cont/testing/TestingArrayHandleMultiplexer.h
./vtkm/cont/testing/TestingCellLocatorRectilinearGrid.h
./vtkm/cont/testing/TestingCellLocatorTwoLevel.h
./vtkm/cont/testing/TestingCellLocatorUniformGrid.h
./vtkm/cont/testing/TestingColorTable.h
./vtkm/cont/testing/TestingComputeRange.h
./vtkm/cont/testing/TestingDeviceAdapter.h
./vtkm/cont/testing/TestingDataSetExplicit.h
./vtkm/cont/testing/TestingDataSetSingleType.h
./vtkm/cont/testing/TestingFancyArrayHandles.h
./vtkm/cont/testing/TestingImplicitFunction.h
./vtkm/cont/testing/TestingPointLocatorSparseGrid.h
./vtkm/cont/testing/TestingRuntimeDeviceConfiguration.h
./vtkm/cont/testing/TestingSerialization.h
./vtkm/cont/testing/TestingVirtualObjectHandle.h
```

### CODE REVIEW

File: var/spack/repos/builtin/packages/vtk-m/package.py
Comment (vicentebolea) [inline var/spack/repos/builtin/packages/vtk-m/package.py]: `testing` seems very ambiguous to me, how about `testing_library` or `testlib`

**Comment by kmorel** [comment]
I'm a little concerned with the name of the spack option. The option is called `testing`, but the option does not actually turn on testing. It just turns on the testing library. I would be confused if I tried to compile vtk-m+testing but none of the tests were compiled. Maybe call the option `testlib`?

Also, you should verify that the `vtkm_cont_testing` library gets compiled and installed. That library exists at least as early as VTK-m 1.7.

File: var/spack/repos/builtin/packages/vtk-m/package.py
Comment (vicentebolea) [inline var/spack/repos/builtin/packages/vtk-m/package.py]: No need to OFF this, the reason why is that when `VTKm_ENABLE_TESTING=ON` it will also `DVTKm_ENABLE_TESTING_LIBRARY=ON`. While we do not have `VTKm_ENABLE_TESTING` in the spack package yet, if we in the future it come to it, it might obscure the mechanism.

**Review by kmorel — DISMISSED** [review]

**Comment by brugger1** [comment]
@kmorel Is `vtkm_cont_testing` the name of a library. I did a grep for it in the install with both `VTKm_ENABLE_TESTING_LIBRARY:BOOL=ON` and `VTKm_ENABLE_TESTING:BOOL=ON` and I couldn't find it.

**Comment by kmorel** [comment]
Yes. It is named `vtkm_cont_testing` (which means it will have a filename like `libvtkm_cont_testing.a` or `libvtkm_cont_testing-1.7.so`), and it should be in the same directory as the other libraries (like `vtkm_cont`).

**Comment by brugger1** [comment]
The library gets built with `VTKm_ENABLE_TESTING_LIBRARY:BOOL=ON`
```
ls lib
cmake			       libvtkm_io-1.7.a
libvtkm_cont-1.7.a	       libvtkm_lodepng-1.7.a
libvtkm_cont_testing-1.7.a     libvtkm_rendering-1.7.a
libvtkm_filter_common-1.7.a    libvtkm_source-1.7.a
libvtkm_filter_contour-1.7.a   libvtkm_worklet-1.7.a
libvtkm_filter_extra-1.7.a     libvtkmdiympi_nompi.a
libvtkm_filter_gradient-1.7.a
```

**Review by vicentebolea — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/vtk-m/package.py
Comment (vicentebolea) [inline var/spack/repos/builtin/packages/vtk-m/package.py]:
```suggestion

```

**Comment by brugger1** [comment]
@vicentebolea I removed the else clause as you suggested and had already changed `testing` to `testlib` in response to Ken's review comments.

**Review by vicentebolea — DISMISSED** [review]
Looks good to me, Thanks!

**Comment by cyrush** [comment]
Hi All, 

Here is how I solved this, however it might not be good for general consumption:

https://github.com/Alpine-DAV/ascent/blob/0a93d376cdee36c672b7aee46776bb12cb227b2f/scripts/uberenv_configs/packages/vtk-m/package.py#L104

If we do use variant, can it default to `ON`?

**Comment by vicentebolea** [comment]
@cyrush thanks for the reference

> If we do use variant, can it default to ON?

I think it is best to default=False in the spack package, thinking on installations that do not need the testing library. It is imperative to keep VTK-m builds as lean as possible due to its lengthy building times.

**Comment by brugger1** [comment]
I notice that Cyrus` update made adding `VTKm_ENABLE_TESTING_LIBRARY:BOOL=ON` dependent on version > 1.7.0, should I do the same?

**Comment by vicentebolea** [comment]
@brugger1 yes you could, before 1.7.0 that flag would be ignored as it was introduced shortly before the 1.7.0 release

File: var/spack/repos/builtin/packages/vtk-m/package.py
Comment (vicentebolea) [inline var/spack/repos/builtin/packages/vtk-m/package.py]:
```suggestion
            if "+testlib" in spec and spec.satisfies('@1.7.0:'):
```

It just hurts the eyes two conditionals which can be merged into one :)

**Review by vicentebolea — CHANGES_REQUESTED** [review]
Before merging please squash all of the commits into a single one.

**Review by vicentebolea — CHANGES_REQUESTED** [review]
Before merging please squash all of the commits into a single one.

**Review by vicentebolea — DISMISSED** [review]
Thanks for the contribution!

**Comment by brugger1** [comment]
You're welcome!

**Comment by alalazo** [comment]
Can you rebase one more time to re-run Gitlab pipelines on a new merge commit?

**Comment by cyrush** [comment]
can we use the spack bot to retrigger?

**Comment by cyrush** [comment]
@spackbot run pipeline

**Comment by alalazo** [comment]
@cyrush Unfortunately not at the moment. `spackbot` will run on the same merge commit that fails :slightly_frowning_face: We are working to add a spackbot command for cases like this, which would trigger a run on a new merge commit.

**Comment by brugger1** [comment]
@alalazo I rebased again and `ci/gitlab-ci` pipeline is still running 18 hours later. My guess is that it will eventually fail. Any other ideas on how to get this to pass?

**Comment by alalazo** [comment]
@brugger1 Last week has been a bit rough with Gitlab pipelines, but tests are finally passing! 🎉

**Review by alalazo — APPROVED** [review]
I see maintainers approved already, so I'll merge this PR.

**Comment by brugger1** [comment]
Thanks a lot! Sorry about my lack of patience.

**Pull request merged**
